dm/master/coordinator/coordinator.go

```diff
@@ -295,36 +310,122 @@ func (c *Coordinator) ObserveWorkers() {
 					c.mu.Lock()
 					if w, ok := c.workers[addr]; ok && name == w.Name() {
 						log.L().Info("worker became online, state: free", zap.String("name", w.Name()), zap.String("address", w.Address()))
-						w.SetStatus(WorkerFree)
+						if source, ok := c.workerToConfigs[addr]; ok {
+							// The worker connect to master before we transfer mysqltask into another worker,
+							// try schedule mysqltask.
+							c.schedule(source)
+							w.SetStatus(WorkerBound)
+						} else {
+							// If this worker has not been in 'workerToConfigs', it means that this worker must have lose connect from master more than 6s,
+							// so the mysql task in worker had stop
+							w.SetStatus(WorkerFree)
+						}
 					} else {
 						// TODO: how to deal with unregister worker
 					}
 					c.mu.Unlock()
-
 				case mvccpb.DELETE:
 					log.L().Info("deletekv", zap.String("kv", string(ev.Kv.Key)))
 					kvs := common.WorkerKeepAliveKeyAdapter.Decode(string(ev.Kv.Key))
 					addr, name := kvs[0], kvs[1]
 					c.mu.Lock()
 					if w, ok := c.workers[addr]; ok && name == w.Name() {
 						log.L().Info("worker became offline, state: closed", zap.String("name", w.Name()), zap.String("address", w.Address()))
-						w.SetStatus(WorkerClosed)
-						cfg := c.configs[addr]
-						delete(c.upstreams, cfg.SourceID)
-						delete(c.configs, addr)
+						if source, ok := c.workerToConfigs[addr]; ok {
+							c.schedule(source)
+						}
 					}
 					c.mu.Unlock()
 				}
 			}
 		case <-c.ctx.Done():
 			log.L().Info("coordinate exict due to context canceled")
 			return
+			//case <-t1.C:
+			//	c.tryRestartMysqlTask()
 		}
 	}
 }
 
-// Schedule schedules a free worker to a upstream.
-// TODO: bind the worker the upstreams and set the status to Bound.
-func (c *Coordinator) Schedule() {
+func (c *Coordinator) schedule(source string) {
+	fmt.Printf("schedule source : %v\n", source)
+	// c.waitingTask <- source
+}
+
+func (c *Coordinator) unschedule(cli *clientv3.Client, cfg config.MysqlConfig) {
+	delete(c.taskConfigs, cfg.SourceID)
+	// TODO: store taskConfigs in ETCD
+}
 
+func (c *Coordinator) tryRestartMysqlTask() {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	scheduleNextLoop := make([]string, 1000)
+	fmt.Printf("tryRestartMysqlTask")
+	hasTaskToSchedule := true
+	for hasTaskToSchedule {
+		select {
+		case source := <-c.waitingTask:
+			if cfg, ok := c.taskConfigs[source]; ok {
+				ret := false
+				if w, ok := c.upstreams[source]; ok {
+					// Try start mysql task at the same worker.
+					c.mu.RUnlock()
```

Comment:
Which one is this `RUnlock` matched `RLock`?

Reply:
The coodinator should unlock mutex when it calls a RPC.